pkg/controller/submariner/submariner_controller.go

```diff
@@ -214,6 +218,40 @@ func (r *ReconcileSubmariner) reconcileDaemonSet(owner metav1.Object, daemonSet
 	return errorutil.WithMessagef(err, "error creating or updating DaemonSet %s/%s", daemonSet.Namespace, daemonSet.Name)
 }
 
+func (r *ReconcileSubmariner) reconcileServiceDiscovery(submariner *submarinerv1alpha1.Submariner, reqLogger logr.Logger, isEnabled bool) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		sdExisting := &submarinerv1alpha1.ServiceDiscovery{}
+		if isEnabled {
+			sd := newServiceDiscoveryCR(submariner)
+			//existing, err := r.submClientSet.SubmarinerV1alpha1().ServiceDiscoveries(submariner.Namespace).Get(serviceDiscoveryCrName, metav1.GetOptions{})
+			err := r.client.Get(context.TODO(), types.NamespacedName{Name: "submariner", Namespace: submariner.Namespace}, sdExisting)
+			if err != nil {
+				err = r.client.Create(context.TODO(), sd)
+			} else {
+				if reflect.DeepEqual(sdExisting, sd) {
+					// No need to issue the update if the resource didn't change
+					return nil
+				}
+				err = r.client.Update(context.TODO(), sd)
+			}
+			if err != nil {
+				reqLogger.Error(err, "Error creating service discovery CR")
+			}
+			return err
+		} else {
+			err := r.client.Get(context.TODO(), types.NamespacedName{Name: "submariner", Namespace: submariner.Namespace}, sdExisting)
+			//Delete if exists
+			if err == nil {
+				err = r.client.Delete(context.TODO(), sdExisting)
+				return err
+			}
```

Comment:
You don't need the extra overhead of first getting the resource - just do the `Delete` and check for and ignore `errors.IsNotFound(err)`.

Reply:
Done.